nbextensions/bulkImportCell/paramsWidget.js

```diff
@@ -0,0 +1,551 @@
+       */
+
+        /*
+       * The input control widget is selected based on these parameters:
+       * - data type - (text, int, float, workspaceObject (ref, name)
+       * - input app - input, select
+       */
+
+
+        // RENDERING
+
+        /*
+      The field widget is a generic wrapper around the input. It serves the following purposes:
+      - intercepts messages in order to display status.
+      appSpec - specifies the whole app
+      parameterSpec - just the segment of the appSpec that specifies this individual parameter
+      value - the initial value of this field
+      closeParameters - a list of "close" parameters, which might be context-dependent. E.g. for an output
+          field, this would be the list of all parameters meant to be output objects, so their names can
+          be cross-validated for uniqueness (optional)
+      */
+        function makeFieldWidget(appSpec, parameterSpec, value, closeParameters) {
+            return paramResolver.loadInputControl(parameterSpec)
+                .then(function (inputWidget) {
+                    let fieldWidget = FieldWidget.make({
+                        inputControlFactory: inputWidget,
+                        showHint: true,
+                        useRowHighight: true,
+                        initialValue: value,
+                        appSpec: appSpec,
+                        parameterSpec: parameterSpec,
+                        workspaceId: workspaceInfo.id,
+                        referenceType: 'name',
+                        paramsChannelName: paramsBus.channelName,
+                        closeParameters: closeParameters
+                    });
+
+                    // Forward all changed parameters to the controller. That is our main job!
+                    fieldWidget.bus.on('changed', function (message) {
+                        paramsBus.send({
+                            parameter: parameterSpec.id,
+                            newValue: message.newValue,
+                            isError: message.isError
+                        }, {
+                            key: {
+                                type: 'parameter-changed',
+                                parameter: parameterSpec.id
+                            }
+                        });
+
+                        paramsBus.emit('parameter-changed', {
+                            parameter: parameterSpec.id,
+                            newValue: message.newValue,
+                            isError: message.isError
+                        });
+                    });
+
+                    fieldWidget.bus.on('touched', function () {
+                        paramsBus.emit('parameter-touched', {
+                            parameter: parameterSpec.id
+                        });
+                    });
+
+
+                    // An input widget may ask for the current model value at any time.
+                    fieldWidget.bus.on('sync', function () {
```

Comment:
sonarcloud is flagging a lot of these as duplicate between here and configure.js.  

i'm still figuring out how these message buses work, do we need them in both places? i understood it that each field widget (or input element) had its own bus so that it could send info about changed state to the top level parameter widget bus. is that not the case?